### PR TITLE
feat(hook): Arns useArnsContract hook

### PR DESCRIPTION
**Pull request by atticusofsparta**
image

### CODE REVIEW

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import { useStateValue } from '../../state/state';
+import { useState, useEffect } from 'react';
+
+const localContractAddress = 'bLAgYxAdX2Ry-nt6aH2ixgvJXbpsEYm28NgJgyqfs-U';
+
+export default function useArNSContract() {
+  const [{ arnsSourceContract }, dispatch] = useStateValue();
+  const [sendingContractState, setSendingContractState] = useState(false);
+
+  try {
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:11]: let's move the try into `dispatchNewContractState` as well as `localProvider`

so use effect will be at the top level, and try/catch will be IN `dispatchNewContractState`

File: src/services/arweave/LocalFilesystemDataProvider.ts [hunk 2]
```diff
@@ -17,6 +17,5 @@ export class LocalFileSystemDataProvider implements SmartweaveContractSource {
     } catch (err) {
       console.log('Unable to load contract state.');
     }
-    return;
```
Comment (dtfiedler) [inline src/services/arweave/LocalFilesystemDataProvider.ts:20]: let's leave this in here. makes the `return` explicit vs. implicit

File: src/types.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 export type ArNSContractState = {
-  records: { [x: string]: ArweaveTransactionId };
+  records: { [x: string]: ArweaveTransactionId } | undefined;
```
Comment (dtfiedler) [inline src/types.ts:2]: why is this needed?

File: src/state/reducer.ts [hunk 4]
```diff
@@ -4,7 +4,7 @@ import type { ArweaveTransactionId, ArNSContractState } from '../types';
 export type Action =
   | { type: 'setWalletAddress'; payload: ArweaveTransactionId }
   | { type: 'setGateway'; payload: string }
-  | { type: 'setArnsContractState'; payload: ArNSContractState };
+  | { type: 'setArnsContractState'; payload: ArNSContractState | unknown };
```
Comment (dtfiedler) [inline src/state/reducer.ts:7]: why is this needed? f it's complaining about the type, we need to make sure we're only dispatching a valid contract state
```suggestion
  | { type: 'setArnsContractState'; payload: ArNSContractState };
```

File: src/App.tsx [hunk 5]
```diff
@@ -7,8 +7,11 @@ import {
 import './index.css';
 import { About, FAQ, Home } from './components/pages';
 import { Layout } from './components/layout';
+import useArNSContract from './hooks/useArNSContract/useArNSContract';
```
Comment (dtfiedler) [inline src/App.tsx:10]: nit: let's create an index.tx in `hooks` - we'll have a lot

File: src/types.ts (hunk 3 above)
Comment (atticusofsparta) [reply src/types.ts:2]: its not, I took it out.

File: src/state/reducer.ts (hunk 4 above)
Comment (atticusofsparta) [reply src/state/reducer.ts:7]: i took this out too

File: src/App.tsx (hunk 5 above)
Comment (atticusofsparta) [reply src/App.tsx:10]: :eyes:

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      }
+      setSendingContractState(true);
+
+      const localState = await localProvider.getContractState(
+        localContractAddress,
+      );
+      if (!localState) {
+        throw Error('ArNS contract state is empty');
+      }
+      console.log(localState);
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:30]:
```suggestion
```

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    }, []);
+    async function dispatchNewContractState(): Promise<void> {
+      try {
+        const localProvider = new LocalFileSystemDataProvider();
+      if (sendingContractState) {
+        return;
+      }
+      setSendingContractState(true);
+
+      const localState = await localProvider.getContractState(
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:24]: rename this to `arnsContractState`

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        localContractAddress,
+      );
+      if (!localState) {
+        throw Error('ArNS contract state is empty');
+      }
+      console.log(localState);
+      dispatch({
+        type: 'setArnsContractState',
+        payload: { records: localState.records },
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:33]:
```suggestion
        payload: arnsContractState,
```

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
+import { LocalFileSystemDataProvider } from '../../services/arweave/LocalFilesystemDataProvider';
+import { useStateValue } from '../../state/state';
+import { useState, useEffect } from 'react';
+
+const localContractAddress = 'bLAgYxAdX2Ry-nt6aH2ixgvJXbpsEYm28NgJgyqfs-U';
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:5]: when hard-coding values, it's common practice to set them as all caps

```suggestion
const ARNS_SOURCE_CONTRACT_ID = 'bLAgYxAdX2Ry-nt6aH2ixgvJXbpsEYm28NgJgyqfs-U';
```

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+export default function useArNSContract() {
+  const [{ arnsSourceContract }, dispatch] = useStateValue();
+  const [sendingContractState, setSendingContractState] = useState(false);
+
+
+
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:12]: nit: `npm run format`

File: src/state/reducer.ts [hunk 11]
```diff
@@ -21,7 +21,7 @@ export const reducer = (state: GlobalState, action: Action): GlobalState => {
     case 'setArnsContractState':
       return {
         ...state,
-        arnsSourceContract: action.payload,
+        arnsSourceContract: action.payload as ArNSContractState,
```
Comment (dtfiedler) [inline src/state/reducer.ts:24]: i don't think this cast is needed with the change from how we're dispatching now (the full state)

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+      });
+
+      setTimeout(() => {
+        // 0.1 second delay before dispatching another calculation to prevent UI jitter
+        setSendingContractState(false);
+      }, 100);
+    }
+    catch (error) {
+      console.log(
+        `error in setting contract state to GlobalState provider >>>`,
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:43]: this is super nit-picky, but I'm a strong proponent for sentences in error messages. 

```suggestion
        `Error in setting contract state.`,
```

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+const ARNS_SOURCE_CONTRACT_ID = 'bLAgYxAdX2Ry-nt6aH2ixgvJXbpsEYm28NgJgyqfs-U';
+
+export default function useArNSContract() {
+  const [{ arnsSourceContract }, dispatch] = useStateValue();
+  const [sendingContractState, setSendingContractState] = useState(false);
+
+  useEffect(() => {
+    dispatchNewContractState();
+  }, []);
+  async function dispatchNewContractState(): Promise<void> {
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:14]: another nit: add space between these 
```suggestion

  async function dispatchNewContractState(): Promise<void> {
```

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 14]
```diff
@@ -2,48 +2,44 @@ import { LocalFileSystemDataProvider } from '../../services/arweave/LocalFilesys
...
-
-    useEffect(() => {
-      dispatchNewContractState();
-    }, []);
-    async function dispatchNewContractState(): Promise<void> {
-      try {
-        const localProvider = new LocalFileSystemDataProvider();
+  useEffect(() => {
+    dispatchNewContractState();
+  }, []);
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:13]:
```suggestion
    }, [arnsSourceContract]);
```

File: src/hooks/useArNSContract/useArNSContract.tsx [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      if (!arnsContractState) {
+        throw Error('ArNS contract state is empty');
+      }
+      dispatch({
+        type: 'setArnsContractState',
+        payload: arnsContractState,
+      });
+
+      setTimeout(() => {
+        // 0.1 second delay before dispatching another calculation to prevent UI jitter
```
Comment (dtfiedler) [inline src/hooks/useArNSContract/useArNSContract.tsx:35]:
```suggestion
        // short delay to prevent another dispatch
```

**Review by dtfiedler — APPROVED** [review]
looks good, nice job!
